### PR TITLE
Add sadad

### PR DESCRIPTION
**Pull request by MehdiSadeghi1996**
it is not tested but can be usefull.
simulated from .net version.

### CODE REVIEW

**Comment by Ja7ad** [comment]
Please use standard pattern for git commit: https://dev.to/jasonh33/git-commit-patterns-5dm7

For example : 
```none
feat: x y z
```

File: providers/sadad/models.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
+package sadad
+
+import (
+	"github.com/GoFarsi/paygap/client"
+	"time"
+)
+
+type MultiplexingType int64
```
Comment (Ja7ad) [inline providers/sadad/models.go:8]: change `MultiplexingType` type to int for enum.

```go
type MultiplexingType int

```

File: providers/sadad/models.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+const (
+	Percentage MultiplexingType = iota
+	Amount
+)
+
+type Sadad struct {
```
Comment (Ja7ad) [inline providers/sadad/models.go:15]: Base struct better use for keep API configuration and Auth to external service, please remove **amount** and etc.

File: providers/sadad/models.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+type PaymentRequest struct {
+	TerminalId    string    `json:"terminal_id"`
+	MerchantId    string    `json:"merchant_id"`
+	Amount        int64     `json:"amount"`
+	OrderId       string    `json:"order_id"`
+	LocalDateTime time.Time `json:"local_date_time"`
+	ReturnUrl     string    `json:"return_url"`
+	SignData      string    `json:"sign_data"`
+	//EnableMultiplexing bool             `json:"enable_multiplexing"`
```
Comment (Ja7ad) [inline providers/sadad/models.go:39]: If this commented field is unrequited, please remove it.

File: providers/sadad/models.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		break
+	case 1:
+		break
+	default:
+		break
+	}
+
+	return true
+}
+func (w MultiplexingType) EnumIndex() int {
```
Comment (Ja7ad) [inline providers/sadad/models.go:89]: please change `MultiplexingType` to int and remove this method.

File: providers/sadad/models.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+type MultiplexingData struct {
+	Type             MultiplexingType       `json:"type"`
+	MultiplexingRows []MultiplexingDataItem `json:"multiplexingRows"`
+}
+type MultiplexingDataItem struct {
+	IbanNumber int32 `json:"iban_number"`
+	Value      int64 `json:"value"`
+}
+
+func (data *MultiplexingData) IsValidated() bool {
```
Comment (Ja7ad) [inline providers/sadad/models.go:51]: This validation required for? explain it please...

File: providers/sadad/sadad.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	"time"
+)
+
+const (
+	baseUrl           = "https://sadad.shaparak.ir"
+	requestPaymentUrl = "/api/v0/Request/PaymentRequest"
+	requestVerifyUrl  = "/api/v0/Advice/Verify"
+)
+
+func New(client client.Transporter, terminalId string, amount int64,
```
Comment (Ja7ad) [inline providers/sadad/sadad.go:23]: This parameters `amount int64, merchantKey string, returnUrl string, merchantId string, purchasePage string` is unrequited, New function is constructor for create sadad object.
Just get configuration and Auth details.

File: providers/sadad/sadad.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	merchantKey string, returnUrl string, merchantId string,
+	purchasePage string, enableMultiplexing bool,
+	multiplexingData *MultiplexingData) (*Sadad, error) {
+
+	if client == nil {
+		return nil, status.ERR_CLIENT_IS_NIL
+	}
+	orderId := string(rand.Int())
```
Comment (Ja7ad) [inline providers/sadad/sadad.go:31]: I think **orderId** is better get from function parameter, just validate it with go validator in struct.

File: providers/sadad/sadad.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	str := []string{terminalId, orderId, string(amount)}
+	joinedString := strings.Join(str, ";")
```
Comment (Ja7ad) [inline providers/sadad/sadad.go:33]: better use fmt.Sprintf for join this fields.

```go
fmt.Sprintf("%s;%s;%d", terminalId, orderId, amount)
```

File: providers/sadad/sadad.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	signedDataAsSadadWay, er := sadad.SigningData(joinedString)
+	if er != nil {
+		panic(er)
+	}
+	sadad.SignData = signedDataAsSadadWay
+
+	return sadad, nil
+}
+
+func (s *Sadad) PayRequest(ctx context.Context) (*PayResultData, error) {
```
Comment (Ja7ad) [inline providers/sadad/sadad.go:60]: In PayRequest better get `PaymentRequest` as parameter.

File: providers/sadad/sadad.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+	return response, nil
+}
+
+func (s *Sadad) SigningData(plainText string) (string, error) {
+
+	key := []byte(s.MerchantKey)
+	plaintext := []byte(plainText)
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		panic(err.Error())
```
Comment (Ja7ad) [inline providers/sadad/sadad.go:131]: Don't use panic in modules and return error.

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: providers/sadad/sadad.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	mode := ecb.NewECBEncrypter(block)
+	ciphertext := make([]byte, len(plaintext))
+	mode.CryptBlocks(ciphertext, plaintext)
```
[the remark this answers is not in the document]
Comment (Ja7ad) [reply providers/sadad/sadad.go:135]: Please check and review this issue.

File: providers/sadad/models.go (hunk 1 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/models.go]: Done.

File: providers/sadad/models.go (hunk 2 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/models.go:15]: Thank you,
Done.

File: providers/sadad/models.go (hunk 3 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/models.go]: Thank you,
it is not unnecessary , it is for co-payments.

File: providers/sadad/models.go (hunk 4 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/models.go]: Thanks, Done.

File: providers/sadad/models.go (hunk 5 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/models.go:69]: Thanks,
This is a validation for user input data when using co-pay.

File: providers/sadad/sadad.go (hunk 7 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/sadad.go]: Thanks.Done.

File: providers/sadad/sadad.go (hunk 8 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/sadad.go]: Thanks, Done.

File: providers/sadad/sadad.go (hunk 9 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/sadad.go]: Thanks. Done.

File: providers/sadad/sadad.go (hunk 10 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/sadad.go:124]: Thanks.Done

File: providers/sadad/sadad.go (hunk 6 above)
Comment (MehdiSadeghi1996) [reply providers/sadad/sadad.go]: Thanks.
Done.

File: providers/sadad/sadad.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (MehdiSadeghi1996) [reply providers/sadad/sadad.go:124]: I implemented according to the this url documentation : 
https://pkg.go.dev/github.com/andreburgaud/crypt2go/ecb

**Pull request merged**